### PR TITLE
[CLOUD-78] external_services: upsert "authorization" field in `Upsert` method

### PR DESCRIPTION
**Pull request by unknwon**
Upon finishing the implementation of CLOUD-78, we couldn't get private repositories being added the user's organization to show up in the user's search results (the user also has access to the repository on the code host).

After the investigation, we discovered that we overlooked `ExternalServicesStore.Upsert` method where we call [`upsertAuthorizationToExternalService`](https://github.com/sourcegraph/sourcegraph/blob/d5ac81bc118795a0b1b759d5d63838d755d4e651/internal/database/external_services.go#L575) in both [`ExternalServicesStore.Create`](https://github.com/sourcegraph/sourcegraph/blob/d5ac81bc118795a0b1b759d5d63838d755d4e651/internal/database/external_services.go#L630-L631) and [`ExternalServicesStore.Update`](https://github.com/sourcegraph/sourcegraph/blob/d5ac81bc118795a0b1b759d5d63838d755d4e651/internal/database/external_services.go#L961-L962), to make sure the special `"authorization"` field exists in every external service config.

The outcome of missing this field caused our permissions syncer to believe that [it does not need to fetch permissions from the code host](https://github.com/sourcegraph/sourcegraph/blob/d5ac81bc118795a0b1b759d5d63838d755d4e651/enterprise/cmd/repo-updater/internal/authz/perms_syncer.go#L458-L460) for the user (because [we do not derive an `authz.Provider` from external services that are not having the `"authorization"` field in their config](https://github.com/sourcegraph/sourcegraph/blob/d5ac81bc118795a0b1b759d5d63838d755d4e651/enterprise/internal/authz/github/authz.go?_pjax=%23js-repo-pjax-container%2C%20div%5Bitemtype%3D%22http%3A%2F%2Fschema.org%2FSoftwareSourceCode%22%5D%20main%2C%20%5Bdata-pjax-container%5D#L86-L89)).

### Security impact

We're totally safe and no private data had leaked thanks to the philosophy that [we never skip authz on dotcom](https://github.com/sourcegraph/sourcegraph/blob/d5ac81bc118795a0b1b759d5d63838d755d4e651/internal/database/external_services.go#L1454).

### Followup

With this discovery, our prod database is in "dirty" state that all external services do not have the "authorization" field. This is fine as-is because we're always on the conservative side and assume no access on dotcom by default.

Technically, we need an out-of-band migration to upsert `"authorization"` field to all existing external services, but given the fact (that I believe) our repo syncing process continuously call `Upsert` method between 2 minutes and 8 hours for all external services, after this PR landed on prod for 8 hours, we will fix the "dirty" state. cc @sourcegraph/repo-management to confirm the "fact" :D

I also added a warning log in this PR for user-added external services that we cannot derive an `authz.Provider` for us to monitoring from logs (and also useful as a general warning).

---

Jira: CLOUD-78

### CODE REVIEW

**Comment by rafax** [comment]
How can we track the number of externalServices without authz set - through a DB query? Log-based metric (in Grafana? somewhere else?)?

**Review by rafax — APPROVED** [review]

File: enterprise/cmd/repo-updater/internal/authz/perms_syncer.go [hunk 1]
```diff
@@ -456,7 +456,13 @@ func (s *PermsSyncer) fetchUserPermsViaExternalServices(ctx context.Context, use
 			return nil, errors.Wrapf(err, "new provider from external service %d", svc.ID)
 		}
 		if provider == nil {
-			// We have no authz provider configured for this external service
+			// NOTE: User code host connection can only be added on sourcegraph.com, and
+			//  authorization is enforced everything, it does not make sense that we cannot
```
Comment (kopancek) [inline enterprise/cmd/repo-updater/internal/authz/perms_syncer.go:460]:
```suggestion
			//  authorization is enforced for everything, it does not make sense that we cannot
```

**Review by kopancek — APPROVED** [review]

**Comment by kopancek** [comment]
> How can we track the number of externalServices without authz set - through a DB query? Log-based metric (in Grafana? somewhere else?)?

I don't think we have a metric for the authorization field, so doing a DB query sounds like the easiest option. It would be great to monitor the count of external services without it after this is deployed to production to verify that we are indeed making the number smaller automatically.

**Comment by unknwon** [comment]
> How can we track the number of externalServices without authz set - through a DB query? Log-based metric (in Grafana? somewhere else?)?

Ideally a DB query would save us as the easiest option but for historical reasons we cannot because:

1. External service config is not standard JSON but JSON with comments, so we can't use Postgres's JSON path syntax for access the value of this column.
2. The entire external service config is encrypted on dotcom, [just happened to have relevant discussion about this in Slack](https://sourcegraph.slack.com/archives/CMBA8F926/p1635945710119600?thread_ts=1635874834.115400&cid=CMBA8F926).

Log-based metric is probably the only viable option, a quick Google search seems indicating [it is available for the GCP logging](https://cloud.google.com/logging/docs/logs-based-metrics). I will follow up on this.

**Comment by unknwon** [comment]
I've pinged `@sourcegraph/repo-management` on Slack, given that the feedback is not blocking, I'm gonna go ahead and merge the PR while waiting.

**Comment by ryanslade** [comment]
> Technically, we need an out-of-band migration to upsert "authorization" field to all existing external services, but given the fact (that I believe) our repo syncing process continuously call Upsert method between 2 minutes and 8 hours for all external services, after this PR landed on prod for 8 hours, we will fix the "dirty" state. cc @sourcegraph/repo-management to confirm the "fact" :D

Correct, after syncing a user owned external service we upsert to record it as synced.

https://github.com/sourcegraph/sourcegraph/blob/main/internal/repos/syncer.go#L509

**Pull request merged**